Quadrature: spec tweaks & Simpson's rule

nshaffer opened the pull request:
### Implementation of `simps` and `simps_weights`

Interfaces and implementations are for 1d real arrays. The included tests exercise the routines for integrands which are (analytically) exactly integrated by Simpson's rule. Known edge cases (length 0, 1, or 6) arrays are explicitly tested as well. There is a lot of new code, and it is not always simple, but hopefully the tests stand as an empirical demonstration of correctness. 

Q: To simplify the implementation, `simps` and `simps_weights` have been made `recursive`. This is only an implementation detail. Am I right to think `pure recursive` does not limit the routines' usability any more than just `pure`, e.g., in `do concurrent` or `where` constructs?

### Changes to `simps` and `simps_weights` optional argument `even`

In the current spec, the optional argument `even` is an integer that is supposed to let the user specify the index at which to insert a Simpson's 3/8 rule so that even-length arrays can be treated. In implementation, this introduced too many corner cases to be worthwhile. In this implementation, `even` is still an integer, but its meaning is different. If it is negative, the 3/8 rule is used for the leftmost abscissas. If it is positive, the 3/8 rule is used for the rightmost abscissas. If it is zero or absent, the result is as if the `even<0` and `even>0` cases were averaged. There are still corner cases and complicated control flow, but it's much simpler than the alternative. The spec has been revised to reflect this new behavior.

### Improved `fypp` style

Looking toward future extensions to complex types and multiple ranks, the `fypp` style has been changed to more closely match that of other heavily preprocessed modules.

### `trapz_weights` example "bugfix"

In the spec for `trapz_weights`, the example had a typo and also used `dot_product` to evaluate the integral. Looking toward implementation for complex integrands, this seems wrong (introduces a complex conjugate), so the example is changed to use `sum` instead.

nshaffer commented:
It looks like the error tolerance is too strict in some of the tests in the CI. Of course they all pass on my machine ;)

jvdp1 commented on src/tests/quadrature/test_simps.f90:
```diff
@@ -0,0 +1,505 @@
+
+    implicit none
+
+    real(sp) :: tol_sp = epsilon(1.0_sp)
+    real(dp) :: tol_dp = epsilon(1.0_dp)
+    real(qp) :: tol_qp = epsilon(1.0_sp)
```
```suggestion
    real(sp) :: tol_sp = 1000 * epsilon(1.0_sp)
    real(dp) :: tol_dp = 1000 * epsilon(1.0_dp)
    real(qp) :: tol_qp = 1000 *epsilon(1.0_sp)
```
As used in other other tests.

jvdp1 commented on src/stdlib_experimental_quadrature.fypp:
```diff
@@ -1,5 +1,6 @@
 #:include "common.fypp"
-
+#:set WEIGHT_FUNS = ["sin", "cos", "pole"]
+#:set QUAD_OK = False
```
How is this variable managed? Through the CMakefile? Manually?
Is it just to avoid the issue with GFortran 9.2?

jvdp1 replied:
I tested it with `QUAD_OK = True` and ifort 19.1, and it went fine.

nshaffer replied:
Ack, all this stuff related to "quad" and such is stuff I'm playing with on the side and should not be in this PR. I will remove it.

jvdp1 commented on src/stdlib_experimental_quadrature.fypp:
If you think it is a bug, did you report it?

nshaffer replied:
gfortran's PDT implementation is fundamentally broken and is known to need a rework. This is experimental code that should not be in the PR anyway.

jvdp1 commented on src/stdlib_experimental_quadrature_simps.fypp:
```diff
@@ -0,0 +1,277 @@
+            integral = dx/3.0_${k1}$*(y(1) + 4*y(2) + y(3))
+        case (4)
+            integral = simps38(y, dx)
+        ! case (5) not needed; handled by default
```
What should happen if I provide a array `y` of size > 6? Currently it will go to the `default case` and all the entries (i.e. more than 5) will be used to compute `integral`.
Instead, should a warning or an error be mentioned? Or is it the intended behaviour?
I usually use `default case` to report such "errors".

nshaffer replied:
The default case handles everything that's not a special case. Here, special cases arise for small arrays, which need special-cased quadrature rules or control flow to do the right thing. It is the nature of these types of quadrature rules to use all the array entries given.

jvdp1 replied:
OK. Thank you for the explanations.

jvdp1 commented on src/stdlib_experimental_quadrature_simps.fypp:
```diff
@@ -0,0 +1,277 @@
+        case (4)
+            integral = simps38(y, x)
+        ! case (6) unneeded?
+        case default
```
Here too: all arrays with a size >4 will end up in the `case default`. Is it intended?

nshaffer replied:
Yep, same comment as above. Though, the questioning "unneeded?" can be changed to a firm "unneeded."

jvdp1 commented on src/stdlib_experimental_quadrature_simps.fypp:
```diff
@@ -0,0 +1,277 @@
+                        ! 3/8 rule on right
+                        integral = simps(y(1:n-3), x(1:n-3)) + simps38(y(n-3:n), x(n-3:n))
+                        return
+                    else
+                        ! fall through
```
```suggestion

```

nshaffer replied:
What is the suggestion here?

jvdp1 replied:
These two lines are not needed and can be removed.

nshaffer replied:
I think it helps make the control flow easier to understand. Compilers will eliminate it anyway.

nncarlson commented on src/stdlib_experimental_quadrature.fypp:
```diff
@@ -14,67 +15,130 @@ module stdlib_experimental_quadrature
+        #:for k1, t1 in REAL_KINDS_TYPES
+        pure module function trapz_x_${k1}$(y, x) result(integral)
+            ${t1}$, dimension(:), intent(in) :: y
+            ${t1}$, dimension(size(y)), intent(in) :: x
```
Is this a private auxiliary function where you know that the actual argument for `x` will be contiguous in practice, or is this a public user-callable function? (I haven't looked carefully at the whole component to know.) If it is the latter, I don't think you want to be declaring `x` as an explicit-shape array, `dimension(size(y))`. This implies F77-style argument passing, and if the actual argument isn't contiguous it will do an implicit copy-in of that argument to a contiguous temporary and pass it instead. Clearly `x` and `y` need to have the same size, but that style of declaration does nothing whatsoever to ensure that that is the case (in fact it hides the true length of the actual argument). This really requires a run-time check or assertion, with `x` declared as an assumed-shape array.

nshaffer replied:
This routine is public by way of the generic `trapz` interface, so your comment is germane. I have no complaints about using assumed-shape declaration here, but I need to educate myself more on array argument passing to understand why it matters. It is a little surprising to learn that the explicit-shape declaration introduces a temporary for non-contiguous actual arguments, whereas the assumed-shape declaration does not.

epagone replied:
What is described is just the implementation detail of (a) specific compiler(s). Although I didn't have the time to check, I strongly doubt that any of such considerations are present in the standard.

nncarlson replied:
It's quite true that this isn't explicitly stated in the standard, however it is implied. The array
```${t1}$, dimension(size(y)), intent(in) :: x``` is simply contiguous (6.5.4) whereas the corresponding actual argument need not be. The only way for that to work is for the array data to be reorganized into contiguous storage.

epagone replied:
Are you sure that the only possible way to match a non-contiguous array with a contiguous one is by implicit copy-in? I don't know the answer.

nncarlson replied:
I think so, but that would be good question for a compiler writer like @pklausler

epagone replied:
I also wonder if allowing a non-contiguous array into the function (declaring `x` as an assumed-shape array) and working on it, would hurt the overall performance more than a copy-in.

klausler replied:
I haven't studied the code in question, and may have misunderstood the question; sorry in advance if that's so.

Arguments that have to be contiguous within the called procedure have to be made so by the caller via allocating, populating, passing, copying back, and deallocating a temporary array when the data being passed might not be contiguous.  Arguments that don't have to be contiguous are cheaper to pass using a "descriptor" or "dope vector", but accesses within the called procedure might require more expensive address calculation, and this cost must be paid even when the actual argument is actually contiguous.  But it's not terribly expensive -- this calculation is invariant in any loop and easy to optimize.

The dummy arguments that don't have to be contiguous are non-pointer non-allocatable "assumed-shape" dummy arrays declared with colons for dimensions.

Some actual argument arrays require temporaries to be created for reasons other than contiguity; if these cases are common, the cost of using an assumed-shape dummy argument isn't buying much other than flexibility to handle the others.  For example, passing the result of an array expression.

My general advice (and this is without knowledge of your specific use case, so weight it lightly) is to use assumed-shape dummy arguments for maximum flexibility, with the usual caveat that you should profile execution time later.  If you want to experiment, note that you can still declare an assumed-shape dummy array to be `CONTIGUOUS`, and doing so will help you determine whether some benefit from contiguity actually matters without losing flexibility.  It depends a lot on your target machine, too; SIMD "vectorization" benefits from contiguity, but it may not matter so much on a GPU.

Hope this helps.

nshaffer replied:
Thanks for the recommendation! I like the approach of making the in-argument assumed shape, and then later benchmarking against making it contiguous as well. 

A related questions is if function results should also be assumed-shape. E.g.,
````fortran
function trapz_weights_dp(x) result(w)
real(dp), dimension(:), intent(in) :: x
real(dp), dimension(size(x)) :: w ! explicit-shape result
! real(dp), dimension(:), allocatable :: w ! assumed-shape result
...
end function trapz_weights_dp
````
The discussion in #114 suggests that allocatable function results are undesirable. Do I understand right that an explicit-shape result is still expected to be the better option?

nncarlson replied:
I'm not sure there was any conclusion in #114 as to which performed better.  The issue probably being whether the array was allocated on the stack or the heap.  I haven't looked to see how you're using this particular function, but personally I've switched to generally avoiding array-valued functions altogether, as I feel this results in a lot of unnecessary allocations/deallocations of memory that passing the result as an argument avoids. I'm not really confident that compiler optimizers are able to eliminate these.

certik replied:
@klausler, thanks for that excellent advice and explanation. @milancurcic we should put this recommendation somewhere in our tutorial section or FAQ, as this question will keep coming up often.

nshaffer replied:
I have just pushed a commit that makes the `x` argument of `trapz_x_*` and `simps_x_*` assumed-shape with a runtime check on its size. Thanks, @nncarlson, for raising the point, and @epagone and @klausler for the discussion.

jvdp1 commented:
Thanks for the answers. I have additional minor comments. In the specs, could you:
* remove "to be implemented" in the titles?
* replace `simps` in `result = simps...` by something like `result = [[stdlib_experimental_quadrature(module):simps(interface)]](....)` for the docs?

nshaffer commented:
> Thanks for the answers. I have additional minor comments. In the specs, could you:
> 
>     * remove "to be implemented" in the titles?
> 
>     * replace `simps` in `result = simps...` by something like `result = [[stdlib_experimental_quadrature(module):simps(interface)]](....)` for the docs?

Sure, no problemo. I see that there are now also links for the docs in `stdlib_experimental_quadrature.fypp` for the `trapz` and `trapz_weights` interfaces. I suppose we want the same for `simps` and `simps_weights`, but I don't know how to get the URL right. Do you know what to do?

jvdp1 commented:
> I suppose we want the same for `simps` and `simps_weights`

It would be good, indeed. You can use a similar link as for `trapz`:
```fortran
        !! Integrates sampled values using trapezoidal rule
        !! ([Specification](../page/specs/stdlib_experimental_quadrature.html#description))	
```
The word `description` must be adapted (e.g., `description_2`)

nshaffer commented:
> > I suppose we want the same for `simps` and `simps_weights`
> 
> It would be good, indeed. You can use a similar link as for `trapz`:
> 
> ```fortran
>         !! Integrates sampled values using trapezoidal rule
>         !! ([Specification](../page/specs/stdlib_experimental_quadrature.html#description))	
> ```
> 
> The word `description` must be adapted (e.g., `description_2`)

Cool beans, made sense after I installed FORD and built the docs. Should be taken care of now.

jvdp1 commented:
> Cool beans, made sense after I installed FORD and built the docs. Should be taken care of now.

Note that you can load it from [here](https://github.com/nshaffer/stdlib/actions/runs/124937976).

jvdp1 commented on src/stdlib_experimental_quadrature_simps.fypp:
```diff
@@ -0,0 +1,278 @@
+        ${t1}$ :: a, b, c
+
+        n = size(y)
+        if (size(x) /= n) error stop "simps: Arguments `x` and `y` must be the same size."
```
I suggest you use the `stdlib` subroutine `error_stop` to avoid issues with older Fortran standards

```suggestion
        if (size(x) /= n) call error_stop("simps: Arguments `x` and `y` must be the same size.")
```
The functions cannot be pure when calling error_stop

nshaffer replied:
I think this is a defect of our `error_stop`. The only functionality it adds over a bare `error stop` statement is to optionally put both a message and an error code. Because that message is written to stderr, purity is lost. At least with a bare `error stop`, we can keep purity with sufficiently recent compilers (e.g., GCC version 8 or newer). I am perfectly happy to pretend pre-f2018 standards do not exist, but I think that does not align with stdlib's current goals. 

The alternative is to revert to declaring `x` to be `dimension(size(y))`. This has the potential performance penalties described in other comments, but at least the compiler/runtime will do the error checking and reporting for us without running afoul of purity rules.

jvdp1 replied:
I don't think reverting to declaring `x` to be `dimension(size(y))` is a good idea, as mentioned in earlier comments.

It has been agreed that Gfortran <=7 were not supported for stdlib. However, recent versions of other compilers may not support the 2018 `error stop` yet. I am perfectly ok to use the 2018 `error stop`, but this will restrict stdlib to very recent compilers.

certik commented:
I would recommend to recommend to use error_stop and open another issue to get rid of it all over stdlib, if desired.

On Sun, Jun 7, 2020, at 10:04 AM, Jeremie Vandenplas wrote:
> 
> 
> *@jvdp1* commented on this pull request.
> 
> In src/stdlib_experimental_quadrature_simps.fypp 
> <https://github.com/fortran-lang/stdlib/pull/209#discussion_r436377661>:
> 
> > +#:for k1, t1 in REAL_KINDS_TYPES
> +
> +    pure recursive module function simps_x_${k1}$(y, x, even) 
> result(integral)
> +        ${t1}$, dimension(:), intent(in) :: y
> +        ${t1}$, dimension(:), intent(in) :: x
> +        integer, intent(in), optional :: even
> +        ${t1}$ :: integral
> +
> +        integer :: i
> +        integer :: n
> +
> +        ${t1}$ :: h1, h2
> +        ${t1}$ :: a, b, c
> +
> +        n = size(y)
> +        if (size(x) /= n) error stop "simps: Arguments `x` and `y` 
> must be the same size."
> I don't think reverting to declaring `x` to be `dimension(size(y))` is 
> a good idea, as mentioned in earlier comments.
> 
> It has been agreed that Gfortran <=7 were not supported for stdlib. 
> However, recent versions of other compilers may not support the 2018 
> `error stop` yet. I am perfectly ok to use the 2018 `error stop`, but 
> this will restrict stdlib to very recent compilers.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/fortran-lang/stdlib/pull/209#discussion_r436377661>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAFAWHJR7OUKKKYDXBNKUDRVO3B3ANCNFSM4NRLGEBA>.
>

nshaffer commented:
> I would recommend to recommend to use error_stop and open another issue to get rid of it all over stdlib, if desired.
> […](#)
> On Sun, Jun 7, 2020, at 10:04 AM, Jeremie Vandenplas wrote: ***@***.**** commented on this pull request. In src/stdlib_experimental_quadrature_simps.fypp <[#209 (comment)](https://github.com/fortran-lang/stdlib/pull/209#discussion_r436377661)>: > +#:for k1, t1 in REAL_KINDS_TYPES + + pure recursive module function simps_x_${k1}$(y, x, even) result(integral) + ${t1}$, dimension(:), intent(in) :: y + ${t1}$, dimension(:), intent(in) :: x + integer, intent(in), optional :: even + ${t1}$ :: integral + + integer :: i + integer :: n + + ${t1}$ :: h1, h2 + ${t1}$ :: a, b, c + + n = size(y) + if (size(x) /= n) error stop "simps: Arguments `x` and `y` must be the same size." I don't think reverting to declaring `x` to be `dimension(size(y))` is a good idea, as mentioned in earlier comments. It has been agreed that Gfortran <=7 were not supported for stdlib. However, recent versions of other compilers may not support the 2018 `error stop` yet. I am perfectly ok to use the 2018 `error stop`, but this will restrict stdlib to very recent compilers. — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#209 (comment)](https://github.com/fortran-lang/stdlib/pull/209#discussion_r436377661)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAFAWHJR7OUKKKYDXBNKUDRVO3B3ANCNFSM4NRLGEBA>.

Ok, done. I've replaced the size check using `error_stop` with a call to `check` and dropped the `pure` attributed from `simps_x_*` and `trapz_x_*`.

jvdp1 commented:
All comments were answered. I'll merge it.
Thank you @nshaffer

certik commented:
@nshaffer do you want to open a new issue for the "error_stop" thing we discussed above?

jvdp1 commented:
> @nshaffer do you want to open a new issue for the "error_stop" thing we discussed above?

Related to #212
